pageserver/src/tenant.rs

```diff
@@ -3405,30 +3410,42 @@ fn run_initdb(
         initdb_bin_path, initdb_target_dir, initdb_lib_dir,
     );
 
-    let initdb_output = Command::new(&initdb_bin_path)
+    let _permit = INIT_DB_SEMAPHORE.acquire().await;
+
+    let initdb_command = tokio::process::Command::new(&initdb_bin_path)
         .args(["-D", initdb_target_dir.as_ref()])
         .args(["-U", &conf.superuser])
         .args(["-E", "utf8"])
         .arg("--no-instructions")
-        // This is only used for a temporary installation that is deleted shortly after,
-        // so no need to fsync it
         .arg("--no-sync")
         .env_clear()
         .env("LD_LIBRARY_PATH", &initdb_lib_dir)
         .env("DYLD_LIBRARY_PATH", &initdb_lib_dir)
         .stdout(Stdio::null())
-        .output()
+        .spawn()
         .with_context(|| {
             format!(
                 "failed to execute {} at target dir {}",
                 initdb_bin_path, initdb_target_dir,
             )
         })?;
-    if !initdb_output.status.success() {
-        bail!(
-            "initdb failed: '{}'",
-            String::from_utf8_lossy(&initdb_output.stderr)
-        );
+
+    let child_id = initdb_command.id();
+
+    tokio::select! {
+        initdb_output = initdb_command.wait_with_output() => {
+            let initdb_output = initdb_output?;
+            if !initdb_output.status.success() {
+                bail!(
+                    "initdb failed: '{}'",
+                    String::from_utf8_lossy(&initdb_output.stderr)
+                );
+            }
+        }
+        _ = cancel.cancelled() => {
+            tokio::process::Command::new("kill").arg("-9").arg(child_id.unwrap().to_string()).output().await?;
+            bail!("initdb was cancelled");
```

Comment:
Regardless of how we solve the killing, we still need to `wait` for the child here before bailing. Otherwise we produce a zombie process.

Reply:
(I guess we could use `kill_on_drop`)?

Reply:
Tokio processes have sigchild based reaping (unless that has been changed to be pidfd based).